### PR TITLE
Add middleware

### PR DESCRIPTION
**Pull request by cesargb**
I hope that this changes will be usefull. I add 2 middelware for greater flexibility. Please, feel free to modify or cancel. Thanks for your time.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -477,54 +477,28 @@ You can use all of the blade directives listed in [using blade directives](#usin
...
-    if ($permission && ! $request->user()->can($permission)) {
-        abort(403);
-    }
-
-    return $next($request);
-}
-```
-
-Don't forget to add the route middleware to `app/Http/Kernel.php` file:
+This package comes with RoleMiddleware and PermissionMiddleware middleware. You can add them inside your app/Http/Kernel.php file.
```
Comment (freekmurze) [inline README.md]: Please wrap all class names and paths in `backticks`

File: src/Middleware/PermissionMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+class PermissionMiddleware
+{
+    public function handle($request, Closure $next, $permission)
+    {
+        if (Auth::guest()) {
+            return $next($request);
+        }
+
+        if ($permission && ! Auth::user()->can($permission)) {
```
Comment (freekmurze) [inline src/Middleware/PermissionMiddleware.php]: It seems odd that the `RoleMiddleware` can accept multiple roles, but this middleware cannot.

Could you add support for multiple permissions?

File: src/Middleware/PermissionMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+use Auth;
+use Closure;
+
+class PermissionMiddleware
+{
+    public function handle($request, Closure $next, $permission)
+    {
+        if (Auth::guest()) {
+            return $next($request);
```
Comment (freekmurze) [inline src/Middleware/PermissionMiddleware.php]: This doesn't seem right. It looks like guest can now access protected page. Could you add a test that makes sure that guest cannot access protected pages and fix this?

File: src/Middleware/RoleMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+use Auth;
+use Closure;
+
+class RoleMiddleware
+{
+    public function handle($request, Closure $next, $role)
+    {
+        if (Auth::guest()) {
+            return $next($request);
```
Comment (freekmurze) [inline src/Middleware/RoleMiddleware.php]: This doesn't seem right. It looks like guest can now access protected page. Could you add a test that makes sure that guest cannot access protected pages and fix this?

File: tests/MiddlewareTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+{
+    public function setUp()
+    {
+        parent::setUp();
+
+        Auth::login($this->testUser);
+    }
+
+    /** @test */
+    public function it_can_access_if_user_has_role()
```
Comment (freekmurze) [inline tests/MiddlewareTest.php]: Could you change `if_user` to `if_the_user` in every added test?

File: tests/MiddlewareTest.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        $request = new Request();
+        $middleware = new RoleMiddleware($this->app);
+
+        $result = $middleware->handle($request, function ($request) {
+            return (new Response())->setContent('<html></html>');
+        }, 'testRole|testrole2');
+    }
+
+    /** @test */
+    public function it_can_access_if_user_has_permiss()
```
Comment (freekmurze) [inline tests/MiddlewareTest.php]: `permiss` --> `permission`

File: src/Middleware/PermissionMiddleware.php (hunk 3 above)
Comment (cesargb) [inline src/Middleware/PermissionMiddleware.php]: @freekmurze I need resolve this  line  ```return redirect($urlOfYourLoginPage);```
My first thought was create a config element to replace $urlOfYourLoginPage. But then why this work not would correspond to middelware Auth?

But when you say this, I understand your position. 

I return my initial idea, add config params to define url to redirect. Perhaps ... ?

```php
if (Auth::guest()) {
    return redirect(config('permission.url_redirect_guest', '/login'));
```

Comment (freekmurze) [inline src/Middleware/PermissionMiddleware.php]: The simplest solution would be to just `abort(403)`. People can do the redirection in the exception handler if they wish.

**Comment by cesargb** [comment]
Hi @freekmurze I was see this commit: https://github.com/spatie/laravel-permission/commit/f590a3f006a85435a1636d5cbee132383c524fd9

I think that, now we can change this code in `PermissionMiddleware.php`:

```php
        if (! Auth::user()->hasAnyPermission(...$permission)) {
            abort(403);
        }
```

for this other:

```php
        if (! Auth::user()->hasAnyPermission($permission)) {
            abort(403);
        }
```

Perhaps, the function `hasAnyRole`, Should I have the same behavior?

**Comment by emielmolenaar** [comment]
Nice changes! :clap:

File: tests/MiddlewareTest.php [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Spatie\Permission\Middleware\RoleMiddleware;
+use Spatie\Permission\Middleware\PermissionMiddleware;
+
+class MiddlewareTest extends TestCase
+{
+    /**
+     * @test
+     * @expectedException Symfony\Component\HttpKernel\Exception\HttpException
```
Comment (freekmurze) [inline tests/MiddlewareTest.php]: I'm not a fan of using the `expectedException` annotation. Could you use the equivalent method instead?

File: tests/MiddlewareTest.php [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+     */
+    public function it_can_not_access_with_middleware_role_if_is_guest()
```
Comment (freekmurze) [inline tests/MiddlewareTest.php]: The name of this test feels a bit off. Could you make this more readable eg `a_guest_cannot_access_a_route_protected_by_the_role_middleware` or something better?

Comment (freekmurze) [inline tests/MiddlewareTest.php]: Do this for all new test methods

File: tests/MiddlewareTest.php (hunk 7 above)
Comment (freekmurze) [inline tests/MiddlewareTest.php]: Remove the annotation on the other new methods as well.

File: tests/MiddlewareTest.php [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        $request = new Request();
+        $middleware = new RoleMiddleware($this->app);
+
+        $result = $middleware->handle($request, function ($request) {
```
Comment (freekmurze) [inline tests/MiddlewareTest.php]: That block

```
$middleware->handle($request, function ($request) {
             return (new Response())->setContent('<html></html>');
         }, 'testRole');
```

is something that is used a lot.

Wrap it into a protected helper method called `testMiddleware` or something better.

```php
function testMiddleware($middleware, $parameter) {
   $middleware->handle($request, function () {
        return (new Response())->setContent('<html></html>');
    }, $parameter)
}
```

**Comment by freekmurze** [comment]
Thanks, I'll fix the remaining nitpicks myself

**Pull request merged**